tiledb/sm/buffer/buffer.h

```diff
@@ -418,6 +438,226 @@ class ConstBuffer : public BufferBase {
+   * same memory as buffer_owner_.
+   */
+  span<const char> buffer_;
+
+ public:
+  /** Marker type for non-owned buffer assign methods. */
+  class NonOwnedMarker {};
+
+  /** Singleton instance of NonOwnedMarker. */
+  static constexpr NonOwnedMarker NonOwned{};
+
+  /**
+   * The type of the allocator used by the buffer. This is required to make the
+   * type allocator-aware.
+   */
+  using allocator_type = decltype(buffer_owner_)::allocator_type;
+
+  /**
+   * Constructor.
+   *
+   * @param alloc Optional allocator for owned buffers.
+   */
+  SerializationBuffer(const allocator_type& alloc = {})
+      : buffer_owner_(alloc)
+      , buffer_(buffer_owner_) {
+  }
+
+  /** Default copy constructor. */
+  SerializationBuffer(const SerializationBuffer& other)
+      : buffer_owner_(other.buffer_owner_)
+      // If we are copying an owned buffer, we must update buffer_ to point to
+      // the new buffer_owner_, otherwise we keep the existing buffer_.
+      , buffer_(other.is_owned() ? buffer_owner_ : other.buffer_) {
+  }
+
+  /** Default copy assignment operator. */
+  SerializationBuffer& operator=(const SerializationBuffer& other) {
+    buffer_owner_ = other.buffer_owner_;
+    buffer_ = other.is_owned() ? buffer_owner_ : other.buffer_;
+    return *this;
+  }
+
+  /** Default move constructor. */
+  // Use of noexcept is important to avoid copies when BufferList is being
+  // resized.
+  SerializationBuffer(SerializationBuffer&& other) noexcept
+      : buffer_owner_(std::move(other.buffer_owner_))
+      // Invalidate the buffer_ from other to avoid use-after-move bugs.
+      , buffer_(std::exchange(other.buffer_, {})) {
+  }
+
+  /** Default move assignment operator. */
+  SerializationBuffer& operator=(SerializationBuffer&& other) {
+    buffer_owner_ = std::move(other.buffer_owner_);
+    buffer_ = std::exchange(other.buffer_, {});
+    return *this;
+  }
+
+  /** Allocator-aware copy constructor. */
+  SerializationBuffer(
+      const SerializationBuffer& other, const allocator_type& alloc)
+      : buffer_owner_(other.buffer_owner_, alloc)
+      , buffer_(other.is_owned() ? buffer_owner_ : other.buffer_) {
+  }
+
+  /** Allocator-aware move constructor. */
+  SerializationBuffer(SerializationBuffer&& other, const allocator_type& alloc)
+      : buffer_owner_(std::move(other.buffer_owner_), alloc)
+      // We can't copy the buffer_ from other because the buffer might be
+      // reallocated if the allocators are different.
+      , buffer_(other.is_owned() ? buffer_owner_ : other.buffer_) {
+    other.buffer_ = {};
+  }
+
+  /**
+   * Constructor for an owned buffer of a given size.
+   *
+   * The data is intended to be modified later with the owned_mutable_span()
+   * function.
+   *
+   * @param size The size of the buffer.
+   * @param alloc Optional allocator for the buffer.
+   */
+  SerializationBuffer(size_t size, const allocator_type& alloc = {})
```

Comment:
I don't think those default parameters are a good idea unless absolutely necessary because people will make mistakes and we won't be tracking some things. We did this for some things in the rtree because adding tracking to everything there was impossible and not having the default allocator would have required tons of code modifications. Why do we need the default parameter here and in other places for the allocator?

Reply:
There are several uses in test code that don't use an allocator and passing one made code more verbose. I understand that we want to make sure no buffer gets untracked (and I just caught some more cases), but it doesn't feel right to me to enforce it by not specifying a default parameter (can't explain why to be honest).

Reply:
If we don't enforce it, it will get missed by some people. It can be tedious to update the test code but I don't see a better way.